akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -78,23 +78,29 @@ import scala.util.control.NonFatal
         if (subSink.isDefined) {
           //delegate to subSink
           subSink.get.pull()
-        } else if (accumulated.size < n) {
+        } else if (accumulated._2 < n) {
           pull(in)
-        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+        } else if (accumulated._2 == n) { //corner case for n = 0, can be handled in FlowOps
           materializeFlow()
         }
       }
 
       override def onDownstreamFinish(cause: Throwable): Unit = {
         if (subSink.isEmpty) {
-          materializeFlow()
+          cause match {
+            case NonFatal(ex) =>
+              downstreamCause = OptionVal.Some(ex)
+            case ex =>
+              super.onDownstreamFinish(ex)
```

Comment:
I think that's not needed. Fatal errors won't be propagated (if they are it's a bug).